Mathlib/Tactic/Widget/StringDiagram.lean

```diff
@@ -0,0 +1,552 @@
+/-
+Copyright (c) 2024 Yuma Mizuno. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yuma Mizuno
+-/
+import ProofWidgets.Component.PenroseDiagram
+import ProofWidgets.Presentation.Expr
+import Mathlib.CategoryTheory.Monoidal.Category
+
+namespace Mathlib.Tactic.Widget.StringDiagram
+
+open Lean Meta Elab
+open CategoryTheory
+
+structure Atom₁ : Type where
+  e : Expr
+
+/-- Expressions for 1-morphisms. -/
+inductive Mor₁ : Type
+  | id : Expr → Mor₁
+  | comp : Mor₁ → Mor₁ → Mor₁
+  | of : Atom₁ → Mor₁
+  deriving Inhabited
+
+def Mor₁.e : Mor₁ → MetaM Expr
+  | .id C => do
+    mkAppOptM ``MonoidalCategoryStruct.tensorUnit #[none, none, C]
+  | .comp f g => do
+    mkAppM ``MonoidalCategoryStruct.tensorObj #[← Mor₁.e f, ← Mor₁.e g]
+  | .of f => return f.e
+
+def Mor₁.toList : Mor₁ → List Expr
+  | .id _ => []
+  | .comp f g => f.toList ++ g.toList
+  | .of f => [f.e]
+
+partial def toMor₁ (e : Expr) : Mor₁ :=
+  match e.getAppFnArgs with
+  | (``MonoidalCategoryStruct.tensorUnit, #[_, _, C]) => Mor₁.id C
+  | (``MonoidalCategoryStruct.tensorObj, #[_, _, _, f, g]) => (toMor₁ f).comp (toMor₁ g)
+  | _ => Mor₁.of ⟨e⟩
+
+/- Expressions for atomic structural 2-morphisms. -/
+inductive StructuralAtom : Type
+  | associator (f g h : Mor₁) : StructuralAtom
+  | associatorInv (f g h : Mor₁) : StructuralAtom
+  | leftUnitor (f : Mor₁) : StructuralAtom
+  | leftUnitorInv (f : Mor₁) : StructuralAtom
+  | rightUnitor (f : Mor₁) : StructuralAtom
+  | rightUnitorInv (f : Mor₁) : StructuralAtom
+  deriving Inhabited
+
+def StructuralAtom.e : StructuralAtom → MetaM Expr
+  | .associator f g h => do
+    mkAppM ``Iso.hom #[← mkAppM ``MonoidalCategoryStruct.associator #[← f.e, ← g.e, ← h.e]]
+  | .associatorInv f g h => do
+    mkAppM ``Iso.inv #[← mkAppM ``MonoidalCategoryStruct.associator #[← f.e, ← g.e, ← h.e]]
+  | .leftUnitor f => do mkAppM ``Iso.hom #[← mkAppM ``MonoidalCategoryStruct.leftUnitor #[← f.e]]
+  | .leftUnitorInv f => do mkAppM ``Iso.inv #[← mkAppM ``MonoidalCategoryStruct.leftUnitor #[← f.e]]
+  | .rightUnitor f => do mkAppM ``Iso.hom #[← mkAppM ``MonoidalCategoryStruct.rightUnitor #[← f.e]]
+  | .rightUnitorInv f => do mkAppM ``Iso.inv #[← mkAppM ``MonoidalCategoryStruct.rightUnitor #[← f.e]]
+
+def structuralAtom? (e : Expr) : Option StructuralAtom := do
```

Comment:
My inclination would be to rewrite this with Qq to make it more robust to metavariables / abbreviations / metadata.